### ISSUE TITLE
Getting Started with New Driver Question

### ISSUE THREAD

**Issue by mjs513** [UAVCAN/libuavcan_kinetis#1]
As I mentioned on the forum we got a new driver going for flexcan for the Teensy boards which are kinetis based (K20, K64 and K66).  However, I would like to update the current clock driver to use PIT timers vs elapsedMillis, it may resolve some issues I am having that I traced to the clock driver.

I have been through the examples but just don't get it - certain things in c++ always get me.  Anyway right now I am doing this to initialize the clock:

`systemClock = &initSystemClock();
// interfaces to systemclock and canDriver
ISystemClock* systemClock;`

where initSystemClock() is defined as:

```
// init system clock interface
ISystemClock& initSystemClock()
{
  return uavcan_teensy::SystemClock::instance();
}
```
but for the life of me I having problems getting it going.  Wonder if you all could help with the correct format.

Thanks
Mike

**Comment by pavel-kirienko** [comment]
> but for the life of me I having problems getting it going

what problems exactly?

**Comment by mjs513** [comment]
Thanks for getting back to me Pavel.  The library is working great so far on the Teensy boards using the associated IDEs. 

Here are some of the error messages (pretty much repeated a bunch of times:
```
C:\Users\CYBERP~1\AppData\Local\Temp\arduino_build_116602\sketch\teensy_uavcan_gnsFix_publish.ino.cpp.o: In function `uavcan_teensy::SystemClock::adjustUtc(uavcan::UtcDuration)':

C:\Users\CyberPalin\Documents\Arduino\libraries\libuavcan_teensy\src/uavcan/uavcan_teensy/clock.hpp:112: undefined reference to `uavcan_teensy::clock::adjustUtc(uavcan::UtcDuration)'

C:\Users\CYBERP~1\AppData\Local\Temp\arduino_build_116602\sketch\teensy_uavcan_gnsFix_publish.ino.cpp.o: In function `uavcan_teensy::SystemClock::getMonotonic() const':

C:\Users\CyberPalin\Documents\Arduino\libraries\libuavcan_teensy\src/uavcan/uavcan_teensy/clock.hpp:117: undefined reference to `uavcan_teensy::clock::getMonotonic()
```
I've a bunch of times to make sure I had all the includes correct and paths identified.  The only thing I could thing of is how he does is systemclock linkage in clock.hpp:
```
/**
 * Adapter for uavcan::ISystemClock.
 */
class SystemClock : public uavcan::ISystemClock
                    , uavcan::Noncopyable
{
    SystemClock() {
    }

    virtual void adjustUtc(uavcan::UtcDuration adjustment) {
        clock::adjustUtc(adjustment);
    }

public:
    virtual uavcan::MonotonicTime getMonotonic() const {
        return clock::getMonotonic();
    }
    virtual uavcan::UtcTime getUtc()             const {
        return clock::getUtc();
    }

/**
 * Calls clock::init() as needed.
 * This function is thread safe.
 */
    static SystemClock& instance();
};
}

```
versus the way it is now:
```
private:
  /*
   * Instance of SystemClock
   */
  static SystemClock self;

  SystemClock() { }

  uavcan::MonotonicTime getMonotonic() const override
  {
    return clock::getMonotonic();
  }

  uavcan::UtcTime getUtc() const override
  {
    return clock::getUtc();
  }

  void adjustUtc(uavcan::UtcDuration adjustment) override
  {
    clock::adjustUtc(adjustment);
  }
};```

**Comment by pavel-kirienko** [comment]
Where are the following functions defined? Seems like their translation unit is not included in the build.

```
C:\Users\CyberPalin\Documents\Arduino\libraries\libuavcan_teensy\src/uavcan/uavcan_teensy/clock.hpp:112: undefined reference to `uavcan_teensy::clock::adjustUtc(uavcan::UtcDuration)'

C:\Users\CyberPalin\Documents\Arduino\libraries\libuavcan_teensy\src/uavcan/uavcan_teensy/clock.hpp:117: undefined reference to `uavcan_teensy::clock::getMonotonic()
```

**Comment by mjs513** [comment]
That's what I thought as well but they are in associated .cpp file unless I just can't see the problem.  Just for reference here are the files I am working with.
[uavcan_clock_new.zip](https://github.com/UAVCAN/libuavcan_kinetis/files/2332954/uavcan_clock_new.zip)

Might as well give you the old clock functions as well, might save some time.
[uavcan_clock_orig.zip](https://github.com/UAVCAN/libuavcan_kinetis/files/2332955/uavcan_clock_orig.zip)

Pavel thanks for the help.  May just have been staring at this so long I can't see the forest because of the trees.

Mike

**Comment by mjs513** [comment]
Hi Pavel,
Isolated the issue but just don't know how to fix.  The functions are defined in the associated clock.cpp file.  However, unless I remove the namespace clock:: from adjustUTC, getMonotonic and getUTC it will always throw those errors.  Just can't figure out a way around it.

If I do that I then get an error message:
```
C:\Users\CYBERP~1\AppData\Local\Temp\arduino_build_493326\sketch\teensy_uavcan_gnsFix_publish.ino.cpp.o: In function `initSystemClock()':

C:\Users\CYBERP~1\AppData\Local\Temp\arduino_build_493326\sketch/phoenix_can_shield.h:45: undefined reference to `uavcan_teensy::SystemClock::instance()
```
and the function initSystemClock() is:
```
// init system clock interface
ISystemClock& initSystemClock()
{   return uavcan_teensy::SystemClock::instance();  }
```
any ideas would be appreciated.

Thanks
Mike

**Comment by pavel-kirienko** [comment]
I will need to have a look at the entire codebase, please.

**Comment by mjs513** [comment]
Not sure how much of the codebase you need but I am attaching the complete lib + complete driver directory + example the example I am testing with.  The only changes to the lib I made were to the driver.  

[uavcan_teensy_driver.zip](https://github.com/UAVCAN/libuavcan_kinetis/files/2339812/uavcan_teensy_driver.zip)

[teensy_uavcan_gnsFix_publish.zip](https://github.com/UAVCAN/libuavcan_kinetis/files/2339818/teensy_uavcan_gnsFix_publish.zip)

[libuavcan_teensy.zip](https://github.com/UAVCAN/libuavcan_kinetis/files/2339819/libuavcan_teensy.zip)

Really appreciate this one since I am really lost on the solution.
Thanks
Mike

**Comment by mjs513** [comment]
Hi Pavel
I took a break from this to kind of clear my head and started to go through the code for kinetis clock driver starting  at the top.  Just a few things I noted right away.  There are no embedded functions for irq_attach, putreg32, up_prioritize_irq, and up_enable_irq.  These seem to be board specific implementations.
Digging into pit timers again the typical kinetis way to set priority and enable the IRQ would be to do something like this:
```
  NVIC_SET_PRIORITY(NVIC_IRQ, 1); /* set interrupt priority */
  NVIC_ENABLE_IRQ(NVIC_IRQ); /* enable interrupt */
``` 
as opposed to: 
```
    // Prioritize and Enable  IRQ
    up_prioritize_irq(TIMX_IRQn, NVIC_SYSH_HIGH_PRIORITY);
    up_enable_irq(TIMX_IRQn);
``` 
which is the way the libuavcan_kinetis is doing it.

So bottom line without some rework I can not use the library for any of the Teensy mcu;s which use a K20, K64 or K66 processor.

I assumed since this was a kinetis library I should be able to use it directly but guess not.

Thanks for the help anyway.
Miek

**Comment by pavel-kirienko** [comment]
Okay. The description of this repo says that this driver is for NuttX-based targets only, so that would make sense. Sorry for the confusion.

**Comment by mjs513** [comment]
Thanks again for you help Pavel.  Think I proved that UAVcan will work on the Teensy and with the Arduino/Teensyduino IDE.  Makes my life easier :).  

Mike
